File: core/x/javaapi/api/src/main/java/viaduct/java/api/resolvers/FieldValue.java
package viaduct.java.api.resolvers;

/**
 * Represents the value of a resolved GraphQL field.
 *
 * <p>Java equivalent of Kotlin's {@code viaduct.api.FieldValue}. Used by batch resolvers to return
 * a per-element success or error for each input context.
 *
 * @param <T> the resolved value type
 */
public sealed interface FieldValue<T> permits FieldValue.Success, FieldValue.Error {

  /** Returns the value on success, or throws the contained exception on error. */
  T get();

  /** Whether this is an error value. */
  boolean isError();

  /** Constructs a {@link FieldValue} that resolved without an error. */
  static <T> FieldValue<T> ofValue(T value) {
    return new Success<>(value);
  }

  /** Constructs a {@link FieldValue} that resolved with the given error. */
  static <T> FieldValue<T> ofError(Exception error) {
    return new Error<>(error);
  }

  record Success<T>(T value) implements FieldValue<T> {
    @Override
    public T get() {
      return value;
    }

    @Override
    public boolean isError() {
      return false;
    }
  }

  @SuppressWarnings("JavaLangClash")
  record Error<T>(Exception error) implements FieldValue<T> {
    @Override
    public T get() {
      sneakyThrow(error);
      throw new IllegalStateException("unreachable");
    }

    @Override
    public boolean isError() {
      return true;
    }

    @SuppressWarnings("unchecked")
    private static <E extends Throwable> void sneakyThrow(Throwable t) throws E {
      throw (E) t;
    }
  }
}

File: core/tenant/api/src/main/kotlin/viaduct/api/FieldValue.kt
package viaduct.api

import viaduct.apiannotations.StableApi

/**
 * Represents the value of a resolved GraphQL object field
 */
@StableApi
sealed interface FieldValue<out T> {
    /**
     * Returns the value on success, or throws an exception for an error value
     */
    fun get(): T

    /**
     * Whether this is an error value or not
     */
    val isError: Boolean

    @StableApi
    companion object {
        /**
         * Constructs a FieldValue that resolved without an error
         */
        fun <T> ofValue(value: T): FieldValue<T> {
            return FieldValueImpl(value)
        }

        /**
         * Constructs a FieldValue that resolved with the [error].
         */
        fun ofError(error: Exception): FieldValue<Nothing> {
            return FieldErrorValueImpl(error)
        }
    }
}

private class FieldValueImpl<T>(
    private val value: T
) : FieldValue<T> {
    override fun get() = value

    override val isError = false
}

private class FieldErrorValueImpl<T>(
    private val error: Exception
) : FieldValue<T> {
    override fun get(): T {
        // TODO (https://app.asana.com/1/150975571430/task/1210815621831967?focus=true): Think through the following:
        // 1. Does this do what we want in terms of having the correct stack trace?
        // 2. How does this interact with TenantResolverException and FrameworkException?
        throw error
    }

    override val isError = true
}
